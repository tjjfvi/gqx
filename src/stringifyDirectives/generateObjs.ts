
import { Obj } from ".";
import { Context } from "..";

export default (ctx: Context) =>
  Object.entries(ctx.objectTypes).map(([type, fields]): Obj => ({
    type,
    props: fields.map(f => ({
      type,
      prop: f.name.value,
      dirs: (f.directives || []).map(d => ({
        type,
        prop: f.name.value,
        dirName: d.name.value,
        dirArgs: d.arguments?.slice(),
      }))
    })),
  }))
