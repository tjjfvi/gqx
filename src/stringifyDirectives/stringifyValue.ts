import { ValueNode } from "graphql";
import indent from "../indent";


const stringifyValue = (val: ValueNode): string =>
  val.kind === "BooleanValue" || val.kind === "StringValue" || val.kind === "EnumValue" ?
    JSON.stringify(val.value) :
    val.kind === "IntValue" || val.kind === "FloatValue" ?
      val.value :
      val.kind === "Variable" || val.kind === "NullValue" ?
        "null" :
        val.kind === "ListValue" ?
          `[\n${val.values.map(v => indent`${stringifyValue(v)},\n`).join("")}]` :
          val.kind === "ObjectValue" ?
            `{\n${val.fields.map(f =>
              indent`${f.name.value}: ${stringifyValue(f.value)},\n`
            ).join("")}}` :
            null

export default stringifyValue;
