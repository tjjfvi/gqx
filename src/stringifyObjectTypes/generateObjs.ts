
import { Context, Obj, Id, Prop } from ".";
import unwrapType from "../unwrapType";

export default (ctx: Context): [Obj[], Id[], Context] => {
  const ids: Id[] = [];
  const objs: Obj[] =
    Object.keys(ctx.objectTypes)
      .sort()
      .map(name => {
        const { fields, unions, implements: impls } = ctx.objectTypes[name];
        const props = fields.map(field => {
          const id: Id = {
            type: name,
            prop: field.name.value,
          };
          ids.push(id);
          const [{ name: { value: typeName } }, wrap,, wrapHKT] = unwrapType(field.type);
          const shallow = !(typeName in ctx.objectTypes);
          let prop: Prop = {
            id,
            type: typeName,
            shallow,
            wrap,
            wrapHKT,
            args: field.arguments?.slice() ?? [],
            directives: field.directives?.slice() ?? [],
            loc: field.loc,
          };
          return prop;
        }).sort((p, q) => p.id.prop > q.id.prop ? 1 : -1);
        const shallowProps = props.filter(p => p.shallow);
        const deepProps = props.filter(p => !p.shallow);
        const obj: Obj = {
          type: name,
          shallowProps,
          deepProps,
          props,
          unions,
          implements: impls,
          isBase: ctx.baseTypes.includes(name),
        };
        return obj;
      });
  return [objs, ids, ctx];
}
