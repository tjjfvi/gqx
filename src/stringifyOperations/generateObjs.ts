
import { Context, Obj, Id, Prop } from ".";
import unwrapType from "../unwrapType";

export default (ctx: Context): [Obj[], Context] => {
  const objs: Obj[] =
    ctx.operations
      .map(([prop, name]) => {
        const { fields } = ctx.objectTypes[name];
        const props = fields.map(field => {
          const id: Id = {
            type: name,
            prop: field.name.value,
          };
          const [{ name: { value: typeName } }, wrap] = unwrapType(field.type);
          let prop: Prop = {
            id,
            type: typeName,
            wrap,
            args: field.arguments?.slice() ?? [],
          };
          return prop;
        }).sort((p, q) => p.id.prop > q.id.prop ? 1 : -1);
        const obj: Obj = {
          type: name,
          prop,
          props,
        };
        return obj;
      })
  return [objs, ctx];
}
