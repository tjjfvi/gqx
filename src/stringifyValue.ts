import { ValueNode } from "graphql";
import { inspect } from "util";
import indent from "./indent";


export const stringifyValue = (val: ValueNode): string =>
  val.kind === "BooleanValue" || val.kind === "StringValue" || val.kind === "EnumValue" ?
    JSON.stringify(val.value) :
    val.kind === "IntValue" || val.kind === "FloatValue" ?
      val.value :
      val.kind === "Variable" || val.kind === "NullValue" ?
        "null" :
        val.kind === "ListValue" ?
          `[\n${val.values.map(v => indent`${stringifyValue(v)},\n`).join("")}]` :
          val.kind === "ObjectValue" ?
            `{\n${val.fields.map(f =>
              indent`${f.name.value}: ${stringifyValue(f.value)},\n`
            ).join("")}}` :
            ((val: never) => {
              throw new Error("Invalid ValueNode " + inspect(val));
            })(val)
