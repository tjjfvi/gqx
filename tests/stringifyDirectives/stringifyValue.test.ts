import generateObjs from "../../src/stringifyDirectives/generateObjs"
import stringifyValue from "../../src/stringifyDirectives/stringifyValue"
import { ValueNode } from "graphql";

describe("stringifyDirectives.gql", () => {
  test.each(
    generateObjs(utils.getOperations(["stringifyDirectives"]))
      .flatMap(x => x.props)
      .flatMap(p => p.dirs)
      .flatMap(d => d.dirArgs)
      .map(f => f.value)
      .map((v): [string, ValueNode] => [v.loc.source.body.slice(v.loc.start, v.loc.end), v])
  )("%s", (_, v) => {
    let val = stringifyValue(v);
    let code = "type $ = " + val;
    expect(code).toParse().toMatchSnapshot();
  })
})

declare global {
  interface Array<T> {
    flatMap<U>(f: (x: T) => U): Array<U extends Array<infer V> ? V : U>;
  }
}
