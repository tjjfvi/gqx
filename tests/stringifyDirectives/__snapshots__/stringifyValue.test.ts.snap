// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`stringifyDirectives.gql "string" 1`] = `"type $ = \\"string\\""`;

exports[`stringifyDirectives.gql "test" 1`] = `"type $ = \\"test\\""`;

exports[`stringifyDirectives.gql ["a","b","c",null] 1`] = `
"type $ = [
  \\"a\\",
  \\"b\\",
  \\"c\\",
  null,
]"
`;

exports[`stringifyDirectives.gql [{x:5},{x:6},{y:[1,2,3]}] 1`] = `
"type $ = [
  {
    x: 5,
  },
  {
    x: 6,
  },
  {
    y: [
      1,
      2,
      3,
    ],
  },
]"
`;

exports[`stringifyDirectives.gql { x: 5, y: [1,2,3], z: { x: 5 } } 1`] = `
"type $ = {
  x: 5,
  y: [
    1,
    2,
    3,
  ],
  z: {
    x: 5,
  },
}"
`;

exports[`stringifyDirectives.gql 12345 1`] = `"type $ = 12345"`;

exports[`stringifyDirectives.gql null 1`] = `"type $ = null"`;

exports[`stringifyDirectives.gql null 2`] = `"type $ = null"`;

exports[`stringifyDirectives.gql null 3`] = `"type $ = null"`;

exports[`stringifyDirectives.gql null 4`] = `"type $ = null"`;

exports[`stringifyDirectives.gql null 5`] = `"type $ = null"`;
